### PR TITLE
Merge spool members through a LazyArray, and window the members a plan trims

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

Follow-up to #1190. When a spool merges several files into one patch, the index route of the streaming merge now builds a `LazyArray` recipe from the member rows and loads it, in place of copying each member into a buffer by hand; both routes finish through one `_assemble`, so they cannot drift. The patch route, which holds trimmed and converted data no source can name, is unchanged.

**Trimmed members are windows.** A member the plan trims used to send the whole merge down the slow patch route, so a `chunk` whose boundaries fall inside files loaded every member as a patch. A trimmed member is now a sample window of its source, taken from the source coordinate's own `select`, so sample-exactness holds by construction. The member rows keep the source's range under private `_{dim}_src_low/_src_high/_src_step` columns (`derived_catalog`), a trim bound is never cast to the coordinate's dtype, and a value selection on the chunked dimension is admitted for the members it did not cut. Everything else — `samples`/`relative` selections, unit-bearing bounds, other dimensions, aux coords, uneven coordinates, unit-converted plans, rows a selection cut, rows with no `origin_id` — takes the patch route, each pinned by a test. Over 100 one-second files, every output of `chunk(time=2.5)`, `chunk(time=0.7)`, `chunk(time=3, overlap=0.5)` and `select(time=...).chunk(time=2.5)` now takes the recipe (none did before) at 1.2–1.4× the speed; whole-file merges, planning and merge-free iteration are at parity with dev (planning at 1e4 and 1e5 files within 3 %; merges of 512 × (8, 3) and 1000 × 4 KB files within 4–6 %, which is the recipe representation itself).

**A file which changed since it was indexed.** A recipe reads exactly the window the index promised, so the load-time shape check which used to catch a rewritten file no longer can. Before a recipe is read, each member source's recorded `(mtime_ns, size_bytes)` is compared with the indexer's own `scan_unit_stats` — a file by its stat, a directory unit by its manifest signature — and any change sends the merge to the patch route. The recorded values ride with the member rows out of the same join, so a plan checks its files against the revision which planned it, carries that in its pickle, asks the index nothing and shares nothing between threads. A member whose record is missing, whose source will not answer a stat, or whose path is not local also takes the patch route. `Spool.from_file` now records its file's stats, measured before and after the scan and recorded only when the two agree; a union of two records which measured one source differently keeps the later record whole, rows and stats together. Same size and mtime with different content is undetectable by design.

**Dtype.** The merged dtype is promoted a member at a time in placement order, as the buffer did (`np.result_type` all at once is not associative: `int16, uint16, float32` is float32 that way and float64 in order), and each member states the dtype it is cast through (`cast_via`), so a chain of promotions rounds where the buffer rounded; the cast is part of the array's `data_id`.

**The index records a coordinate's full dtype.** `CoordSummary` no longer truncates `datetime64[ms]` to `datetime64`, so a datetime coordinate rebuilt from the index comes back at the unit the file stored, on both routes (dev's index route returned `ns`). `INDEX_VERSION` 22 → 23; older indexes rebuild. `CoordSummary.dtype` is user-visible and now carries the unit.

**Accepted.** A patch merged on the index route has `history == ()`, as it already did for whole-file merges on dev; this PR extends that to trimmed merges. History is being deprecated in favour of a provenance tracker.

**Found on dev, not fixed here.** The patch route selects twice and a float32 coordinate's `select` is not idempotent, so a chunk boundary at a member edge drops one sample (`[12, 13, 14, 16, ...]`); the recipe route and the in-memory spool keep it. `chunk(70ms)` then `chunk(None, tolerance=inf)` over non-ns datetime coords raises `TypeError: unhashable type` from `_conformed` returning a 0-d array. `chunk().select(samples=True).chunk(tolerance=inf)` raises `CoordError`. An exact-grid coordinate with a `samples=True` selection exports to xarray one sample short. A `chunk(fill_value=...)` output's `data_id` differs run to run.

**Verification.** Seeded differential harnesses compare the recipe route, the forced patch route and `origin/dev` on archives with integer, float and datetime coordinates at several units, the chunked dimension on every axis, transposed members, multi-patch files, gaps, overlaps, selections, unions of single-file spools and of two indexes, and in-memory patches mixed in: about 2,000 operation sequences and 7,000 output patches, no difference in data, dtype, shape, dims, coordinates or attrs apart from the float32 case above. `scripts/differential_check.py --ref origin/dev`: 900 calls identical. `cast_via` was checked against 69,904 simulated promotion chains and 69,888 real streaming merges.

**Review.** Five blind reviewers, then four Codex reviews, each with executed reproductions, over four fix rounds: an integer coordinate windowed off its grid; a guard added to the shared `coord_from_row` which broke `io.to_xarray()`; an eager stat map built on every `chunk()`; a lazily cached stat record which was stale after `update()`, raced, depended on reopening a deleted index and rescanned the table; a trimmed member admitted with no `origin_id`; all-at-once promotion; NULL stats admitting the recipe; a union pairing one revision's rows with another's stats; nanosecond mtimes lost through float64; chained casts not rounding; a file rewritten between scan and stat; a cast constant sharing an id with its uncast self. Each has a test which failed before its fix and a hand mutant which the tests kill.

Checks: full suite 13917 passed, 131 skipped, 2 xfailed; doctests 268 passed; `pre-commit run --all` clean; no uncovered new lines.

## Changelog

- changed: a spool merge whose members the index fully describes reads only the samples it needs, so chunk boundaries inside files no longer load whole patches; a file which changed since it was indexed is noticed and read as a patch.
- changed: the spool index records a coordinate's full dtype, unit included, so a datetime coordinate rebuilt from the index keeps the unit its file stored; the index version is now 23 and older indexes rebuild.
- changed: merging patches of mixed dtypes promotes a member at a time, in order, on every route.

## Checklist

I have:

- [x] filled in the Changelog section above (see `dascore/docs/contributing/general_guidelines.qmd`).

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [x] documented the new feature with docstrings and/or appropriate doc page.
- [x] included tests. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [x] added the "ready_for_review" tag once the PR is ready to be reviewed.

## Summary by CodeRabbit

- **New Features**
  - Added controlled intermediate and final data-type casting when combining arrays.
  - Indexed data assembly now supports efficient windowed reconstruction while preserving coordinate data types and units.

- **Bug Fixes**
  - Corrected handling of time-based coordinate units.
  - Improved detection of changed files and directory contents to prevent stale indexed results.
  - Kept source metadata consistent when files are updated.

- **Tests**
  - Expanded coverage for casting, coordinate precision, indexed assembly, stale sources, and selection edge cases.

### CODE REVIEW

File: dascore/io/index/indexer.py [hunk 1]
```diff
@@ -43,6 +44,52 @@
...
+    source made of many files, answers with its manifest fingerprint,
+    since its own stat moves for neither a rewritten nor a resized
+    member. A source the filesystem will not answer for states nothing.
+    """
+    try:
+        status = os.stat(path)
+    except OSError:
+        return None, None
+    if stat.S_ISDIR(status.st_mode):
+        return _directory_signature(Path(path))
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/io/index/indexer.py]: Fixed: the directory walk now sits inside the same `except OSError` as the root stat, so a member which vanishes between the listing and its stat leaves the unit with no stats, and the merge takes the patch route. Pinned by `TestScanUnitStats::test_a_member_lost_during_the_walk_states_nothing`.

File: dascore/core/coords.py [hunk 2]
```diff
@@ -278,7 +278,7 @@ def get_correct_dtype_cast_values(cls, data: Any) -> Any:
             data = dict(data)
             min_val = data["min"]
             dtype = _get_dtype(min_val, data.get("dtype"))
-            data["dtype"] = str(dtype).split("[")[0]
+            data["dtype"] = str(dtype)
             for name in ["min", "max", "step"]:
                 val = data.get(name)
                 data[name] = ensure_consistent_dtype(val, name, dtype)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/coords.py:284]: Fixed: `to_coord` casts the summary's nanosecond scalars to the unit the summary records before building the range, so `coord.to_summary().to_coord()` gives back the same dtype and the same `data_id` for s/ms/us/ns datetimes and for timedeltas. Pinned by `TestSummaryRoundTripUnit`. (On dev the summary had no unit to give, so this was unfixable there.)

File: dascore/core/lazy_array.py [hunk 3]
```diff
@@ -53,7 +53,7 @@
 SOURCE_FIELDS = ("base_uri", "path", "format", "version")
 
 # The dictionary encoded member columns; `filled` is a plain bool array.
-MEMBER_FIELDS = ("source", "key", "origin_id", "dtype", "value")
+MEMBER_FIELDS = ("source", "key", "origin_id", "dtype", "cast", "value")
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/lazy_array.py:56]: Fixed: `from_frame` treats an absent `cast` column as no cast, so a frame written before members stated one rebuilds the same array with the same id. Pinned by `TestCastVia::test_a_frame_with_no_cast_column_reads_back`.

File: dascore/core/coords.py [hunk 4]
```diff
@@ -330,6 +330,13 @@ def to_coord(self) -> CoordRange:
             start, stop = self.max, self.min + step
         else:
             start, stop = self.min, self.max + step
+        # The scalars are kept at nanoseconds; the coord takes the recorded unit.
+        if self.dtype and (dtype := np.dtype(self.dtype)).kind in "mM":
+            start, stop = (
+                np.asarray(start).astype(dtype)[()],
+                np.asarray(stop).astype(dtype)[()],
+            )
+            step = np.asarray(step).astype(f"m8[{np.datetime_data(dtype)[0]}]")[()]
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/coords.py]: Fixed: the step is now cast with the unit's count as well (`m8[10ms]`), so `datetime64[10ms]` round-trips with its dtype and `data_id`. Pinned by `TestSummaryRoundTripUnit::test_a_counted_unit_keeps_its_count`, which fails without the fix.

File: dascore/utils/patch_assembly.py [hunk 5]
```diff
@@ -586,53 +677,120 @@ def _merge_patches_streaming(self, joined, df_dict_list, merge_dim, samples):
...
+            return None
+        try:
+            data = recipe.load()
+        except (InvalidFiberIOError, ParameterError):
+            # the resource did not hold the array its row described: a
+            # different shape, or fewer axes than the windows name
+            return None
+        coords = [meta.coords for meta in metas]
+        attrs = [meta.attrs for meta in metas]
+        return self._assemble(data, dims, merge_dim, coords, attrs)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/utils/patch_assembly.py:719]: Declining. The window between the stat check and the read is a few microseconds, and a check after `load()` would leave exactly the same window on its other side — a file can change after any check. What can be detected is already caught: a source whose shape or dtype no longer matches what the index promised makes the loader raise, and the merge falls back to the patch route. A rewrite which keeps the same size and mtime inside that window is undetectable by design, which the PR body states. The same stat-then-read sequence is what the indexer itself relies on.
